feat: 서버 기반 대중교통 경로 공유 및 실시간 이동 상태 추적

benef88 opened the pull request:
📋 Summary

  참가자들이 선택한 대중교통 경로를 Firebase에 저장하고 실시간으로 공유하며, Polyline 기반 정확한 ETA 계산과 이동 상태 추적 기능을 추가했습니다.

  ✨ Main Features

  1. 서버 기반 경로 정보 저장 및 공유

  - ParticipantRoute에 selectedRoute 필드 추가: 전체 대중교통 경로 정보(버스/지하철 노선, 환승 정보 등) 저장
  - Firebase DTO 추가: RouteDto, SubPathDto, LaneDto, StationDto, GraphicCoordinateDto
  - SaveSelectedRouteUseCase: 경로 선택 시 서버에 저장
  - 지도 화면에서 서버 경로 렌더링: 선택된 참가자의 경로를 지도에 표시

  2. Polyline 기반 정확한 ETA 계산

  - CalculateRouteProgressUseCase: TMAP graphicData 기반 정확한 경로 추적
  - 현재 구간(SubPath) 및 진행률 계산: 사용자가 어느 구간을 이동 중인지 실시간 파악
  - 남은 거리/시간 계산: 실제 경로 polyline 기반 정확한 ETA 제공

  3. 이동 상태 자동 추적

  - MovementStatus enum 추가: NOT_STARTED, READY, IN_PROGRESS, ARRIVING_SOON, ARRIVED
  - CalculateMovementStatusUseCase: 위치, 경로, 목적지 기반 자동 상태 판단
  - 실시간 상태 업데이트: 위치 변경 시 자동으로 이동 상태 갱신

  4. 세션 복구 및 데이터 유지

  - 기존 참가자 데이터 보존: 앱 재시작 시 location, route, movementStatus 유지
  - name/profileColor만 업데이트: 세션 복구 시 필수 정보만 갱신
  - 경로 정보 영속성: 앱 종료 후에도 선택한 경로 유지

  5. UI/UX 개선

  - 경로 선택 버튼 상태 관리: 경로가 있으면 FAB으로 축소, 없으면 전체 너비 버튼
  - 서버 경로 기반 버튼 표시: 로컬 경로뿐만 아니라 서버에서 받아온 경로도 반영
  - 출발지 선택 시 검색 결과 초기화: 이전 검색 결과 자동 클리어

  6. 코드 품질 개선

  - 레이어 분리: 하드코딩된 값을 PlaceConstants로 분리
  - 유틸리티 분리: LocationValidator를 별도 클래스로 추출
  - 비즈니스 로직 분리: ViewModel → UseCase로 로직 이동

  🔧 Technical Details

  새로 추가된 파일:
  - CalculateMovementStatusUseCase.kt: 이동 상태 계산
  - CalculateRouteProgressUseCase.kt: 경로 진행률 계산
  - SaveSelectedRouteUseCase.kt: 경로 저장
  - UpdateLocationAndEtaUseCase.kt: 위치 및 ETA 통합 업데이트
  - MovementStatus.kt: 이동 상태 enum
  - PlaceConstants.kt: 장소 관련 상수
  - LocationValidator.kt: 위치 비교 유틸리티

  주요 변경 파일:
  - ParticipantRoute.kt: selectedRoute, 경로 진행 정보 필드 추가
  - RoomDataSource.kt: updateParticipantRoute에 selectedRoute 파라미터 추가
  - FirebaseRoomDataSource.kt: 참가자 데이터 업데이트 로직 개선
  - MapScreen.kt: 서버 경로 렌더링, 버튼 상태 관리

  📊 Stats

  - 27 files changed
  - +1,302 additions, -180 deletions

  🧪 Test Plan

  - 경로 선택 후 Firebase에 selectedRoute 저장 확인
  - 앱 재시작 후 경로 정보 유지 확인
  - 지도에서 선택된 참가자의 경로 렌더링 확인
  - 이동 중 ETA 및 상태 자동 업데이트 확인
  - 경로 선택 버튼 축소/확장 동작 확인

  📝 Notes

  - Firebase에 저장되는 경로 정보에는 SubPath별 graphicData가 포함되어 있어 정확한 경로 표시 가능
  - 위치 업데이트 정책: 25m 이상 이동 또는 30초 경과 시 업데이트
  - 세션 복구 시 기존 데이터를 덮어쓰지 않아 경로/위치 정보가 안전하게 보존됨

benef88 committed:
feat: Polyline 기반 정확한 ETA 계산 로직 추가

- ParticipantRoute에 경로 진행률 추적 필드 추가
- CalculateRouteProgressUseCase 작성 (polyline 기반 진행률 계산)
- UpdateLocationAndEtaUseCase 작성 (위치+ETA 통합 업데이트)
- BroadcastLocationUseCase에 Route 파라미터 추가
- MapViewModel에서 선택된 경로를 UseCase에 전달

benef88 committed:
feat: 참가자 이동 상태 추적 기능 추가

- MovementStatus enum 추가 (NOT_STARTED, READY, IN_PROGRESS, ARRIVING_SOON, ARRIVED)
- CalculateMovementStatusUseCase 구현 (거리 기반 상태 계산)
- UpdateLocationAndEtaUseCase에서 위치 업데이트 시 이동 상태 자동 계산
- Route 모델에 출발지/도착지 좌표 추가
- Participant 모델에 movementStatus 필드 추가
- Firebase 연동 및 실시간 동기화
- UI에 이동 상태별 색상 표시

기술적 세부사항:
- 도착: 목적지로부터 100m 이내
- 곧 도착: 목적지로부터 500m 이내
- 출발 전: 출발지로부터 200m 이내
- 이동 중: 위 조건 외 모든 경우
- Haversine 공식 사용한 정확한 거리 계산

benef88 committed:
feat: 경로 선택 버튼 UI/UX 개선

- 경로 선택 전: 하단 카드뷰와 너비 맞춘 전체 너비 버튼
- 경로 선택 후: FAB 아이콘 버튼으로 최소화 및 우측 배치
- 지도 가시성 향상을 위한 조건부 UI 렌더링

benef88 committed:
feat: 출발지 선택 시 이전 검색 결과 자동 초기화 기능 추가

- SelectPlaceRoute에 clearPreviousResult 파라미터 추가
- LaunchedEffect로 화면 진입 시 이전 검색 결과 초기화
- SelectPlaceViewModel에 clearSearchResult() 메서드 구현
- 출발지 선택 화면(SelectStartPlace)에서 clearPreviousResult=true 설정
- ClearSelectedPlaceUseCase 의존성 주입 추가

이를 통해 출발지 선택 화면에서 이전 모임 장소 검색 결과가
자동으로 초기화되어 사용자 혼란을 방지합니다.

benef88 committed:
refactor: 레이어 분리 개선 - 하드코딩 제거 및 유틸리티 분리

주요 변경사항:
1. 하드코딩된 값을 상수로 분리 (PlaceConstants)
   - DEFAULT_LATITUDE/DEFAULT_LONGITUDE: 기본 위치 좌표
   - REVERSE_GEOCODING_DEBOUNCE_MS: 디바운스 시간 (500ms)
   - LOCATION_COMPARISON_PRECISION: 위치 비교 정밀도 (0.000001)
   - 외부 라이브러리 의존성 제거 (LatLng 대신 원시 타입 사용)

2. 위치 검증 로직을 유틸리티로 분리
   - LocationValidator 클래스 생성 (feature/place/util)
   - isSameLocation() 메서드를 ViewModel에서 제거
   - 재사용 가능한 유틸리티로 분리

3. UI 레이어의 위치 결정 로직을 ViewModel로 이동
   - selectedPosition StateFlow 추가
   - SelectPlaceScreen의 remember 로직 제거
   - 단방향 데이터 흐름 강화

4. ViewModel 개선
   - LocationValidator 의존성 주입
   - PlaceConstants 사용으로 매직 넘버 제거
   - 코드 가독성 향상

개선 효과:
- 하드코딩된 값 제거로 유지보수성 향상
- 위치 검증 로직 재사용 가능
- UI와 비즈니스 로직 분리 강화
- 클린 아키텍처 원칙 준수 (core 모듈의 외부 의존성 제거)

benef88 committed:
feat: 서버 기반 경로 정보 저장 및 지도 렌더링

- ParticipantRoute에 selectedRoute 필드 추가하여 전체 경로 정보 저장
- Firebase용 DTO 추가 (RouteDto, SubPathDto, LaneDto, StationDto, GraphicCoordinateDto)
- RoomMapper에 Route ↔ RouteDto 변환 로직 추가
- SaveSelectedRouteUseCase 생성하여 경로 선택 시 서버에 저장
- SelectRouteViewModel/Screen에서 경로 확정 시 서버에 저장
- UpdateLocationAndEtaUseCase에서 selectedRoute 필드 포함
- updateParticipantRoute에 selectedRoute 파라미터 추가
- FirebaseRoomDataSource에서 selectedRoute를 Firebase에 저장
- MapScreen에서 선택된 참가자의 서버 경로 데이터로 지도에 그리기

이제 참가자들이 선택한 대중교통 경로(버스/지하철 정보 포함)가
Firebase에 저장되고, 다른 참가자들이 지도에서 확인할 수 있습니다.

benef88 committed:
fix: 세션 복구 시 기존 참가자 데이터 유지

- addParticipant에서 기존 참가자 확인 로직 추가
- 기존 참가자는 데이터를 덮어쓰지 않고 유지
- 신규 참가자만 새로 생성
- 앱 재시작 후에도 경로/위치 정보가 유지됨

benef88 committed:
fix: 서버에서 받아온 경로가 있을 때도 경로 선택 버튼 축소

- MapScreen에서 경로 버튼 축소 조건에 서버 경로 확인 추가
- 로컬 경로 또는 서버 경로가 있으면 버튼이 FAB으로 축소
- 앱 재시작 후에도 경로가 있으면 버튼이 축소된 상태로 표시

Bongpal-dev commented on core/data-api/src/main/kotlin/com/teammanduk/adego/core/data_api/model/ParticipantDto.kt:
```diff
@@ -21,5 +22,69 @@ data class ParticipantRouteDto(
     val distanceInMeters: Int = 0,
-    val updatedAt: Long = 0L
+    val updatedAt: Long = 0L,
+    val selectedRoute: RouteDto? = null // 선택한 대중교통 경로
```
DTO 정리가 필요할것 같습니다.

Firebasse에 저장되는 부분은 DataSource에서 처리되어야 하는 부분이라고 생각하는데 현재 DTO는 구현에 대한 모든 것을 표현하려고 하는것 같습니다.

DTO는 통합으로 처리하고 각 Impl에서 상세모델로 전송하는게 맞지 않을까 싶습니다.

Kirby-JH replied:
경로 세부 정보는 룸db 저장 및 인코딩한 polyline 추가하는 방식으로 변경 리뷰 지적해주신 대로 DTO 아키텍처를 완전히 개선했습니다.

  변경사항 (커밋 b045efd):
  - core/data-api에 RouteDto 및 관련 DTO 추가 (@Serializable 적용)
  - SelectedRouteDataSource를 Domain Model 대신 DTO 사용하도록 변경
  - RouteMapper.kt 추가하여 Route ↔ RouteDto 양방향 변환
  - 각 계층의 명확한 책임 분리:
    - Domain Layer: Route (Domain Model만 사용)
    - Data Layer: Route ↔ RouteDto 변환 담당
    - Data-API Layer: RouteDto 정의
    - Local Layer: RouteDto ↔ Entity 변환

  추가 개선 (커밋 6bc6742):
  - Domain Model에서 @Serializable 어노테이션 완전 제거
  - core/model 모듈에서 kotlinx-serialization 의존성 제거
  - 사용하지 않는 RouteTypeConverters.kt 삭제
  - Clean Architecture 원칙 완벽 준수

Bongpal-dev commented on core/common/src/main/kotlin/com/teammanduk/adego/core/common/constants/PlaceConstants.kt:
```diff
@@ -0,0 +1,30 @@
+     * 기본 위치 - 경도 (현재 위치를 가져올 수 없을 때 사용)
+     * 부산 만덕동 좌표
+     */
+    const val DEFAULT_LONGITUDE = 129.0758
```
만덕동 좌표 굿

Kirby-JH replied:
👍

Bongpal-dev commented on core/data-api/src/main/kotlin/com/teammanduk/adego/core/data_api/model/ParticipantDto.kt:
```diff
@@ -21,5 +22,69 @@ data class ParticipantRouteDto(
+ */
+data class GraphicCoordinateDto(
+    val latitude: Double = 0.0,
+    val longitude: Double = 0.0
```
그래픽 좌표는 어디서 사용되는 정보인가요??

Kirby-JH replied:
CalculateRouteProgressUseCase에서 Polyline 기반 정확한 진행률 계산에 사용됩니다.

  사용 위치:
  - core/domain/src/main/kotlin/com/teammanduk/adego/core/domain/usecase/CalculateRouteProgressUseCase.k    
  t:111-122
  - core/domain/src/main/kotlin/com/teammanduk/adego/core/domain/usecase/CalculateRouteProgressUseCase.k    
  t:181-218

  사용 목적:
  1. 사용자 위치에서 가장 가까운 SubPath 찾기 (polyline 좌표들과 거리 비교)
  2. 해당 SubPath 내에서 정확한 진행률 계산 (polyline 포인트 순회하며 누적 거리 계산)

  TMAP API에서 제공하는 실제 경로 좌표 데이터(graphicData)를 활용하여, 단순 직선 거리가 아닌 실제 도로를    
   따라 이동한 거리를 정확하게 계산합니다.

Bongpal-dev commented on core/data/src/main/kotlin/com/teammanduk/adego/core/data/mapper/RoomMapper.kt:
```diff
@@ -93,6 +112,128 @@ fun ParticipantRoute.toDto(): ParticipantRouteDto {
+    )
+}
+
+fun Lane.toDto(): LaneDto {
```
앞서 DTO가 정리가 안됐던 것처럼 Model도 각 DTO와 매핑되다 보니 전체적으로 이를 정리하고 통합해서 서비스 모델을 설정할 필요가 있을것 같습니다.

Kirby-JH replied:
각 Model이 대응하는 DTO를 갖도록 완전히 정리했습니다:
  - Route ↔ RouteDto                                                                                        
  - SubPath ↔ SubPathDto                                                                                    
  - GraphicCoordinate ↔ GraphicCoordinateDto                                                                
  - Station ↔ StationDto                                                                                    
  - Lane ↔ LaneDto                                                                                          
  - TrafficType ↔ TrafficTypeDto

Bongpal-dev commented on core/domain/src/main/kotlin/com/teammanduk/adego/core/domain/usecase/CalculateMovementStatusUseCase.kt:
```diff
@@ -0,0 +1,103 @@
+
+        // 도착: 목적지로부터 100m 이내
+        if (distanceToDestination <= 100) {
+            return MovementStatus.ARRIVED
```
이런 계산들은 Status내부에서 남은거리를 전닯다아서 설정하도록 하는건 어떨까요?

Kirby-JH replied:
거리 임계값을 MovementStatus enum으로 이동했습니다.

  변경 전:
```kt
  // CalculateMovementStatusUseCase.kt                                                                      
  if (distanceToDestination <= 100) {  // 하드코딩                                                          
      return MovementStatus.ARRIVED
  }
  if (distanceToDestination <= 500) {  // 하드코딩                                                          
      return MovementStatus.ARRIVING_SOON
  }
  if (distanceFromStart <= 200) {  // 하드코딩                                                              
      return MovementStatus.READY
  }
```
  변경 후:
```kt
  // MovementStatus.kt                                                                                      
  enum class MovementStatus(val distanceThresholdMeters: Int?) {
      NOT_STARTED(null),
      READY(200),
      IN_PROGRESS(null),
      ARRIVING_SOON(500),
      ARRIVED(100);

      companion object {
          const val ARRIVAL_THRESHOLD = 100                                                                 
          const val ARRIVING_SOON_THRESHOLD = 500                                                           
          const val READY_THRESHOLD = 200                                                                   
      }
  }

  // CalculateMovementStatusUseCase.kt                                                                      
  if (distanceToDestination <= MovementStatus.ARRIVAL_THRESHOLD) {
      return MovementStatus.ARRIVED
  }
  if (distanceToDestination <= MovementStatus.ARRIVING_SOON_THRESHOLD) {
      return MovementStatus.ARRIVING_SOON
  }
  if (distanceFromStart <= MovementStatus.READY_THRESHOLD) {
      return MovementStatus.READY
  }
```
  개선 효과:
  - 거리 임계값이 MovementStatus와 함께 관리되어 응집도 향상
  - 임계값 변경 시 한 곳만 수정하면 됨
  - enum 각 값의 의미가 더 명확해짐

Bongpal-dev commented on core/domain/src/main/kotlin/com/teammanduk/adego/core/domain/usecase/UpdateLocationAndEtaUseCase.kt:
```diff
@@ -0,0 +1,174 @@
+        val userId = userRepository.getCurrentUserId()
+
+        // 위치 업데이트
+        roomRepository.updateMyLocation(userId, location)
```
업데이트 정책 통일을 위해서 해당 UseCase로 이동하는건 어떤가요?

Kirby-JH replied:
해당 파일 삭제 및 UpdateLocationUseCase로 통합완료

Bongpal-dev commented on core/domain/src/main/kotlin/com/teammanduk/adego/core/domain/usecase/CalculateMovementStatusUseCase.kt:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        // 그 외: 이동 중
+        return MovementStatus.IN_PROGRESS
```
지금 .MovementStatus가 거리를 기반으로 도착여부를 나타내는걸로 보이는데 Movement가 맞을지, 그리고 200m이상일때는 멈춰있어도 이동중으로 나오는건지 궁금하네요!

Kirby-JH replied:
현재 정책은 임의로 지정한것이라 추후 수정 및 개선방안 검토 예정입니다.

Bongpal-dev commented on feature/route/src/main/kotlin/com/teammanduk/adego/feature/route/SelectRouteViewModel.kt:
```diff
@@ -94,6 +100,28 @@ class SelectRouteViewModel @Inject constructor(
+            Log.d("SelectRouteViewModel", "선택한 경로를 서버에 저장 중...")
+            val result = saveSelectedRouteUseCase(
+                userId = userId,
+                selectedRoute = selectedRoute
```
userId랑 selectRoute를 ui에서 받아와서 다시 넘기고 있습니다.

Kirby-JH replied:
userId는 제거, selectedRoute는 해당하는 경로의 인덱스를 ui에서만 관리하고 있기 때문에 뷰모델 내부에서 처리하고 있습니다.

Bongpal-dev commented on feature/route/src/main/kotlin/com/teammanduk/adego/feature/route/SelectRouteScreen.kt:
```diff
@@ -46,6 +46,10 @@ internal fun SelectRouteRoute(
+
+            // 선택한 경로 정보를 콜백으로 전달
             uiState.selectedRouteIndex?.let { index ->
                 uiState.routes.getOrNull(index)?.let { route ->
```
경로에 대한 정보가 Result로 전달되어서 viewModel에 ui state를 업데이트한다는걸 결국 해당 정보를 Repository에서도 가지고 있다는건데, 왜 ui state에 있는 경로정보를 다시 전달해서 사용하는 건가요?

Kirby-JH replied:
```kt
 // 현재: UI State의 route를 UseCase로 전달    
                                                           
  val selectedRoute = _uiState.value.routes.getOrNull(selectedIndex)
  saveSelectedRouteUseCase(selectedRoute)

  // 질문: Repository가 이미 경로 목록을 반환했는데,                                                       
  // 다시 같은 데이터를 받아서 저장하는 것이 적절한가?                                                     
```
현재 방식이 더 명시적이고 상태 관리 책임이 명확하다고 생각합니다.
  - ViewModel: UI 상태 관리
  - Repository: 데이터 소스 관리 (캐싱 불필요)
  - 명시적 데이터 흐름 (디버깅 용이)

  하지만 더 나은 패턴이 있다면 논의 필요합니다!

Bongpal-dev commented on feature/place/src/main/kotlin/com/teammanduk/adego/feature/place/util/LocationValidator.kt:
```diff
@@ -0,0 +1,24 @@
+        val latDiff = kotlin.math.abs(pos1.latitude - pos2.latitude)
+        val lngDiff = kotlin.math.abs(pos1.longitude - pos2.longitude)
+        return latDiff < PlaceConstants.LOCATION_COMPARISON_PRECISION &&
+               lngDiff < PlaceConstants.LOCATION_COMPARISON_PRECISION
```
실질적으로 위치가 같은지는 ui에서 처리할게 아니라 data쪽에서 처리를 하는게 맞지않을까요???

Kirby-JH replied:
- 커밋: a093cc3                                                                                                                                                                                                                   
  - LocationValidator → core/domain/util/LocationUtils 이동
  - Domain Layer로 도메인 로직 재배치

Bongpal-dev commented on feature/map/src/main/kotlin/com/teammanduk/adego/feature/map/model/ParticipantUiModel.kt:
```diff
@@ -24,13 +26,27 @@ fun Participant.toUiModel(): ParticipantUiModel = ParticipantUiModel(
     name = this.name,
     color = this.getParticipantColor(),
     status = if (route?.eta != null) "도착중" else "대기중",
```
이부분은 MobementStatus랑 겹쳐지는것 같네요

Kirby-JH replied:
- ParticipantUiModel에서 status 필드 삭제
  - movementStatus만 사용하여 더 정확한 이동 상태 표시
  - UI에서 실제로 movementStatus만 사용하고 있었던 불필요한 중복 제거

Bongpal-dev commented on feature/map/src/main/kotlin/com/teammanduk/adego/feature/map/component/ParticipantCardPager.kt:
```diff
@@ -123,18 +123,30 @@ private fun ParticipantCard(
-                        style = AdegoTheme.typography.bodyLarge,
-                        color = Color(0xFF9E9E9E)
+                        color = when (participant.movementStatus) {
+                            "도착" -> Color(0xFF4CAF50)
```
enum을 만들고, string으로 변환한 다음 그걸로 조건을 만드는 이유가 있나요?

Kirby-JH replied:
- 커밋: 71d44a5                                                                                                                                                                                                                   
  - enum 직접 비교로 변경, 타입 안전성 확보

Bongpal-dev commented on feature/map/src/main/kotlin/com/teammanduk/adego/feature/map/MapViewModel.kt:
```diff
@@ -154,7 +154,10 @@ class MapViewModel @Inject constructor(
+            val selectedRoute = _uiState.value.searchedRoutes
+                .getOrNull(_uiState.value.selectedRouteIndex ?: -1)
+
+            broadcastLocation(selectedRoute)
```
ui 데이터 재탕 2222

benef88 committed:
refactor: UpdateLocationUseCase에 ETA 업데이트 기능 통합

- UpdateLocationAndEtaUseCase 제거 (불필요한 중복)
- UpdateLocationUseCase를 확장하여 선택적으로 Route 파라미터 받도록 수정
- 중복 코드 제거 (shouldUpdate, calculateDistance)
- BroadcastLocationUseCase에서 통합된 UpdateLocationUseCase 사용
- 하위 호환성 유지: invoke(location)과 invoke(location, route) 모두 지원

Bongpal-dev commented on core/domain/src/main/kotlin/com/teammanduk/adego/core/domain/usecase/UpdateLocationAndEtaUseCase.kt:
```diff
@@ -0,0 +1,174 @@
+    /**
+     * 남은 시간을 "15분" 형식으로 포맷
+     */
+    private fun formatEta(remainingTimeInSeconds: Int): String {
```
포맷 변경은 common같은데 extentions로 놓고 ui에서 하는게 맞지 않을까요?

ui에 표현되는 정보를 usecase에서 다룰 이유가...

Kirby-JH replied:
변경사항 (커밋 145d228):
  - 시간 포맷팅 로직을 UseCase에서 core/ui/util/FormatUtils.kt로 이동
  - formatTime(), formatDistance() 등 공통 확장 함수로 통합
  - feature/map과 feature/route에서 공통 사용
  - Domain Layer는 순수 비즈니스 로직만 처리

Bongpal-dev commented on core/domain/src/main/kotlin/com/teammanduk/adego/core/domain/usecase/CalculateRouteProgressUseCase.kt:
```diff
@@ -0,0 +1,250 @@
+    private fun findNearestSubPath(
+        userLocation: ParticipantLocation,
+        subPaths: List<SubPath>
+    ): Triple<Int, SubPath, Double>? {
```
트리플을 사용하는것보다 data class를 사용하는거 어떨까요

Kirby-JH replied:
- CalculateRouteProgressUseCase의 findNearestSubPath에서 Triple 제거
  - NearestSubPath data class 추가 (index, subPath, distance)
  - it.third → it.distance로 가독성 향상
  - 각 필드의 의미가 명확해져 코드 이해도 개선

benef88 committed:
feat: Room DB + Firebase polyline 하이브리드 경로 저장 방식 구현

Clean Architecture 원칙을 준수하며 Firebase 비용을 절감하는 경로 저장 시스템 구현

[아키텍처]
Domain → Repository → DataSource → Storage (Room DB / Firebase)

[핵심 변경사항]
1. 경로 저장 방식
   - Room DB: 상세 경로 데이터 (SubPath, graphicData 등)
   - Firebase: 인코딩된 polyline만 저장

2. 역할 분리
   - SaveSelectedRouteUseCase: 경로 선택 시 1회 실행
     * Room DB 저장 + Polyline 인코딩 + Firebase 업데이트
   - UpdateLocationUseCase: 위치 업데이트마다 실행
     * Room DB 자동 조회 + ETA 계산 + Firebase 업데이트

3. Polyline 처리
   - PolylineEncoder: Google Polyline Algorithm 구현
     * encode(): Route → polyline 문자열
     * decode(): polyline 문자열 → 좌표 리스트
   - MapScreen: polyline 디코딩하여 지도에 표시

[상세 구현]
Domain Layer:
- SelectedRouteRepository 인터페이스
- PolylineEncoder 유틸리티 (인코딩/디코딩)
- SaveSelectedRouteUseCase, UpdateLocationUseCase 수정
- BroadcastLocationUseCase 파라미터 제거

Data Layer:
- SelectedRouteRepositoryImpl
- DataModule DI 바인딩

Data-API Layer:
- SelectedRouteDataSource 인터페이스
- RoomDataSource: 진행률 필드 추가

Local Layer:
- Room DB 인프라 (Entity, DAO, Database, TypeConverter)
- Kotlinx Serialization으로 SubPath 직렬화

Model Layer:
- Route 관련 모델에 @Serializable 추가
- ParticipantRoute에서 selectedRoute 제거

Firebase:
- ParticipantRouteDto 간소화 (selectedRoute/RouteDto 제거)
- 진행률 필드 추가

Feature Layer:
- MapScreen: polyline 디코딩으로 경로 표시
- MapViewModel: selectedRoute 파라미터 제거
- MapNavigation: getSelectedRoute 제거
- MainNavHost: selectedRoute 상태 관리 제거

Build:
- Room 2.6.1, Kotlinx Serialization 의존성 추가

[이점]
- Clean Architecture 준수 (의존성 방향 올바름)
- Firebase 비용 대폭 절감 (상세 데이터 로컬 저장)
- 실시간 ETA/거리/진행률 유지
- 다른 참가자 경로도 polyline으로 시각화 가능
- 경로 데이터 자동 조회 (상태 관리 불필요)

benef88 committed:
refactor: 시간 저장 단위를 초로 통일하고 불필요한 변환 제거

변경사항:
- Route.totalTime, SubPath.sectionTime을 초 단위로 변경
- TMAP API 응답(초) → Route(분) → ParticipantRoute(초) 불필요한 변환 제거
- 모든 저장 레이어(Route, ParticipantRoute, Firebase)에서 초 단위 사용
- UI에서만 formatTime() 함수로 초→분 포맷팅

추가 개선:
- formatTime() 함수 중복 제거: TimeFormatter.kt로 통합
- ParticipantUiModel에서 etaInSeconds, distanceInMeters 처리
- formatDistance() 함수 추가

장점:
- 데이터 일관성: API 응답부터 저장까지 모두 초 단위
- 변환 오류 방지: 초→분→초 불필요한 변환 제거
- 코드 중복 제거: 공통 함수 통합
- 정밀도 유지: 초 단위로 저장하여 정확한 계산 가능
- UI 유연성: 분/시간 단위 표시는 UI에서만 처리

benef88 committed:
fix: 시간 단위 변환 버그 수정 및 코드 정리

주요 버그 수정:
- CalculateRouteProgressUseCase에서 sectionTime * 60 제거 (88분이 88시간으로 표시되던 버그)
- sectionTime은 이미 초 단위이므로 분→초 변환 불필요

추가 개선:
- 포맷팅 함수를 core/ui/util/FormatUtils.kt로 통합 (DRY 원칙)
- feature/map과 feature/route에서 공통 사용
- 프리뷰용 샘플 데이터를 초 단위로 수정
- 불필요한 .toDouble() 제거 (자동 타입 변환)

benef88 committed:
refactor: Clean Architecture 완성 - RouteDto 도입으로 완벽한 계층 분리

변경사항:
- core/data-api에 RouteDto 및 관련 DTO 추가 (@Serializable)
- SelectedRouteDataSource를 RouteDto 사용하도록 변경
- Route ↔ RouteDto Mapper 추가 (RouteMapper.kt)
- SelectedRouteRepositoryImpl에서 Domain Model ↔ DTO 변환
- SelectedRouteDataSourceImpl에서 DTO ↔ Entity 변환

아키텍처 개선:
- Domain Layer: Route (Domain Model만 사용)
- Data Layer: Route ↔ RouteDto 변환 담당
- Data API Layer: RouteDto 정의
- Local Layer: RouteDto ↔ Entity 변환 담당

장점:
- 완벽한 계층 분리 (각 계층이 자신의 책임만 수행)
- 의존성 역전 원칙 준수 (Domain이 Data를 모름)
- 일관성 (다른 DataSource들과 동일한 DTO 패턴)
- 유지보수성 향상 (변경 영향 범위 최소화)

benef88 committed:
refactor: Domain Model 레이어 정리 및 불필요한 의존성 제거

- Domain Model에서 @Serializable 어노테이션 제거 (Route, SubPath, GraphicCoordinate, Station, Lane)
- core/model 모듈에서 kotlinx-serialization 의존성 제거
- 사용하지 않는 RouteTypeConverters.kt 삭제
- AdeGoDatabase에서 @TypeConverters 어노테이션 제거
- Domain 레이어가 Data 레이어에 의존하지 않도록 Clean Architecture 원칙 준수

이제 직렬화는 Data 레이어의 RouteDto에서만 처리됩니다.

benef88 committed:
refactor: MovementStatus enum에 거리 임계값 추가

- MovementStatus에 distanceThresholdMeters 프로퍼티 추가
- companion object에 ARRIVAL_THRESHOLD, ARRIVING_SOON_THRESHOLD, READY_THRESHOLD 상수 정의
- CalculateMovementStatusUseCase의 하드코딩된 거리 값을 enum 상수로 변경
- 거리 임계값 관리를 UseCase에서 Domain Model로 이동하여 응집도 향상

benef88 committed:
refactor: ParticipantUiModel의 중복 status 필드 제거

- status 필드 삭제 (route 유무로 "도착중"/"대기중" 판단하던 필드)
- movementStatus만 사용하여 더 정확한 이동 상태 표시
- UI에서 실제로 movementStatus만 사용하고 있어 불필요한 중복 제거

benef88 committed:
refactor: Triple 대신 명확한 data class 사용

- CalculateRouteProgressUseCase의 findNearestSubPath에서 Triple 제거
- NearestSubPath data class 추가 (index, subPath, distance)
- it.third 대신 it.distance로 가독성 향상
- 각 필드의 의미가 명확해져 코드 이해도 개선

benef88 committed:
refactor: UI 계층 분리 개선 - enum 직접 사용 및 RouteUiModel 도입

1. ParticipantUiModel의 movementStatus를 String에서 MovementStatus enum으로 변경
   - ParticipantCardPager에서 String 비교 대신 enum 직접 비교
   - 타입 안전성 향상 및 컴파일 타임 에러 검출
   - toKoreanString()을 public 확장 함수로 변경하여 UI에서 표시 시 사용

2. RouteUiModel 도입으로 UI 데이터 재사용 문제 해결
   - RouteUiModel 및 SubPathUiModel 생성
   - Route ↔ RouteUiModel 양방향 mapper 추가
   - MapUiState의 searchedRoutes를 RouteUiModel로 변경
   - MapViewModel에서 Domain Model → UiModel 변환
   - MapDialogs의 RouteListDialog를 RouteUiModel 사용하도록 변경

개선 효과:
- UI State가 UI 전용 모델만 포함하여 계층 분리 명확화
- Domain Model과 UI Model의 책임 분리
- 다른 모델(Room, Participant)과 일관된 구조
- PR 리뷰 피드백 완전 반영

benef88 committed:
refactor: 위치 비교 로직을 Domain Layer로 이동

- LocationValidator를 core/domain/util/LocationUtils로 이동
- UI Layer의 비즈니스 로직을 Domain Layer로 재배치
- SelectPlaceViewModel에서 LocationUtils 사용하도록 수정
- DI 불필요 (object singleton 방식)

개선 효과:
- 계층 분리 명확화 (도메인 로직이 Domain Layer에 위치)
- 재사용성 향상 (모든 feature 모듈에서 접근 가능)
- 테스트 용이성 개선 (Domain 모델만 있으면 테스트 가능)
- Place 확장 함수 제공으로 편의성 증가

Bongpal-dev commented on core/domain/src/main/kotlin/com/teammanduk/adego/core/domain/usecase/UpdateLocationUseCase.kt:
```diff
@@ -2,35 +2,105 @@ package com.teammanduk.adego.core.domain.usecase
+                    updatedAt = System.currentTimeMillis()
+                )
+
+                roomRepository.updateMyRoute(userId, participantRoute)
```
내부에서 update를 일괄로 처리하는게 어떨까요?
